README.md
<div align="center">
    <a href="https://github.com/ker0x/twig-image-placeholder-extension/actions" title="Build">
        <img src="https://img.shields.io/github/workflow/status/ker0x/twig-image-placeholder-extension/ci?style=for-the-badge" alt="Build">
    </a>
    <a href="https://codecov.io/gh/ker0x/twig-image-placeholder-extension/" title="Coverage">
        <img src="https://img.shields.io/codecov/c/gh/ker0x/twig-image-placeholder-extension?style=for-the-badge" alt="Coverage">
    </a>
    <a href="https://php.net" title="PHP Version">
        <img src="https://img.shields.io/badge/php-%3E%3D%208.1-8892BF.svg?style=for-the-badge" alt="PHP Version">
    </a>
    <a href="https://packagist.org/packages/kerox/twig-image-placeholder-extension" title="Downloads">
        <img src="https://img.shields.io/packagist/dt/kerox/twig-image-placeholder-extension.svg?style=for-the-badge" alt="Downloads">
    </a>
    <a href="https://packagist.org/packages/kerox/twig-image-placeholder-extension" title="Latest Stable Version">
        <img src="https://img.shields.io/packagist/v/kerox/twig-image-placeholder-extension.svg?style=for-the-badge" alt="Latest Stable Version">
    </a>
    <a href="https://packagist.org/packages/kerox/twig-image-placeholder-extension" title="License">
        <img src="https://img.shields.io/packagist/l/kerox/twig-image-placeholder-extension.svg?style=for-the-badge" alt="License">
    </a>
</div>

# Twig Image Placeholder Extension

A Twig extension to generate images placeholder

> Inspired by [simple-svg-placeholder](https://github.com/cloudfour/simple-svg-placeholder)

## Installation

You can install the extension using Composer:

```
composer require kerox/twig-image-placeholder-extension
```

#### Symfony

If you are using [Symfony Flex](https://flex.symfony.com/) you're done.

Otherwise, add the following lines to your `services.yaml`

```yaml
services:
  ...

  Kerox\TwigImagePlaceholder\SvgPlaceholderExtension: null
```

#### Standalone

If you use Twig as standalone, then you need to add the extension manually

```php
$extension = new \Kerox\TwigImagePlaceholder\SvgPlaceholderExtension();

$twig = new \Twig\Environment($loader);
$twig->addExtension($extension);
```

## Usage

The following functions are available

* [SvgExtension](./src/SvgPlaceholderExtension.php)
    * svg_placeholder()
    * svg_placeholder_uri()

## Examples

#### Default

![](./examples/default.svg?sanitize=true)

```twig
{{ svg_placeholder() }}
```

#### Size

![](./examples/size.svg?sanitize=true)

```twig
{{ svg_placeholder(150, 150) }}
```

#### Colors

![](./examples/colors.svg?sanitize=true)

```twig
{{ svg_placeholder(300, 150, {
    bgColor: '#0f1c3f' , 
    textColor: '#7fdbff'
}) }}
```

#### Text

![](./examples/text.svg?sanitize=true)

```twig
{{ svg_placeholder(300, 150, {
   text: 'Foo Bar'
}) }}
```

Or Without:

![](./examples/notext.svg?sanitize=true)

```twig
{{ svg_placeholder(300, 150, {
   text: false
}) }}
```

#### Data URI

```twig
<img src="{{ svg_placeholder_uri() }}">
```

will output

```html
<img src="data:image/svg+xml;charset=UTF-8,%3Csvg%20xmlns%3D%22http%3A%2F%2Fwww.w3.org%2F2000%2Fsvg%22%20width%3D%22300%22%20height%3D%22150%22%20viewBox%3D%220%200%20300%20150%22%3E%3Crect%20fill%3D%22%23ddd%22%20width%3D%22300%22%20height%3D%22150%22%3E%3C%2Frect%3E%3Ctext%20fill%3D%22rgba%280%2C0%2C0%2C0.5%29%22%20font-family%3D%22sans-serif%22%20font-size%3D%2230%22%20font-weight%3D%22bold%22%20dy%3D%2210.5%22%20x%3D%2250%25%22%20y%3D%2250%25%22%20text-anchor%3D%22middle%22%3E300x150%3C%2Ftext%3E%3C%2Fsvg%3E">
```

## Options Reference

| Option     | Type                    | Default                                            | Description |
| ---------- | ----------------------- | -------------------------------------------------- | ----------- |
| text       | `string`,`null`,`false` | Placeholder dimensions                             | The text to display. If set to `false`, element `text` will not be added to the svg.  |
| fontFamily | `string`                | `sans-serif`                                       | The font to use for the text. For data URIs, this needs to be a system-installed font. |
| fontWeight | `string`                | `bold`                                             | |
| fontSize   | `float`                 | 20% of the shortest image dimension, rounded down. | |
| dy         | `float`                 | 35% of the `fontSize`                              | Adjustment applied to the [dy attribute](https://developer.mozilla.org/en-US/docs/Web/SVG/Attribute/dy) of the text element so it will appear vertically centered. |
| bgColor    | `string`                | `#ddd`                                             | The background color of the image. Defaults to `#ddd`. |
| textColor  | `string`                | `rgba(0,0,0,0.5)`                                  | The color of the text. For transparency, use an `rgba` or `hsla` color value. |
| class      | `string`                |                                                    | If provide, add a `class` attribute to the `svg` element.


CHANGELOG.md
# CHANGELOG

The Twig Image Placeholder Extension library follows [SemVer](http://semver.org/).

## 2.x

**Changelog** (since [`1.1.2`](https://github.com/ker0x/twig-image-placeholder-extension/compare/1.1.2...2.0.0))

- 2.0.0 (2022-10)
    - Drop support for PHP < 8.1
    - Drop support for Twig < 3.0
    - Update dependencies

## 1.x

**Changelog** (since [`1.1.1`](https://github.com/ker0x/twig-image-placeholder-extension/compare/1.1.1...1.1.2))

- 1.1.2 (2021-01)
    - Update .gitattributes
    
**Changelog** (since [`1.1.0`](https://github.com/ker0x/twig-image-placeholder-extension/compare/1.1.0...1.1.1))

- 1.1.1 (2020-12)
    - Allow to be installed on PHP 8.0
    - Add PHP 8.0 to GH workflows
    - Remove `coverage-upload` step in workflows

**Changelog** (since [`1.0.0`](https://github.com/ker0x/twig-image-placeholder-extension/compare/1.0.0...1.1.0))

- 1.1.0 (2020-12)
    - Add option `class`
    - Option `text` can be set to `false` or `null`
    - Add GitHub workflows


tests/SvgExtensionTest.php
<?php

declare(strict_types=1);

namespace Kerox\TwigImagePlaceholder\Tests;

use Kerox\TwigImagePlaceholder\SvgPlaceholderExtension;
use PHPUnit\Framework\TestCase;
use Twig\TwigFunction;

class SvgExtensionTest extends TestCase
{
    protected SvgPlaceholderExtension $extension;

    protected function setUp(): void
    {
        $this->extension = new SvgPlaceholderExtension();
    }

    public function testGetFunctions(): void
    {
        self::assertContainsOnlyInstancesOf(TwigFunction::class, $this->extension->getFunctions());
    }

    public function testGetSvgPlaceholder(): void
    {
        self::assertSame('<svg xmlns="http://www.w3.org/2000/svg" width="300" height="150" viewBox="0 0 300 150"><rect fill="#ddd" width="300" height="150"></rect><text fill="rgba(0,0,0,0.5)" font-family="sans-serif" font-size="30" font-weight="bold" dy="10.5" x="50%" y="50%" text-anchor="middle">300x150</text></svg>', $this->extension->getSvgPlaceholder());
    }

    public function testGetSvgPlaceholderWithOptions(): void
    {
        $options = [
            'bgColor' => '#0F1C3F',
            'textColor' => '#7FDBFF',
            'text' => 'foo bar',
            'class' => 'thumbnail',
        ];

        self::assertSame('<svg xmlns="http://www.w3.org/2000/svg" width="500" height="250" viewBox="0 0 500 250" class="thumbnail"><rect fill="#0F1C3F" width="500" height="250"></rect><text fill="#7FDBFF" font-family="sans-serif" font-size="50" font-weight="bold" dy="17.5" x="50%" y="50%" text-anchor="middle">foo bar</text></svg>', $this->extension->getSvgPlaceholder(500, 250, $options));
    }

    public function testGetSvgPlaceholderWithoutText(): void
    {
        $options = [
            'text' => false,
        ];

        self::assertSame('<svg xmlns="http://www.w3.org/2000/svg" width="500" height="250" viewBox="0 0 500 250"><rect fill="#ddd" width="500" height="250"></rect></svg>', $this->extension->getSvgPlaceholder(500, 250, $options));
    }

    public function testGetSvgPlaceholderUri(): void
    {
        self::assertSame('data:image/svg+xml;charset=UTF-8,%3Csvg%20xmlns%3D%22http%3A%2F%2Fwww.w3.org%2F2000%2Fsvg%22%20width%3D%22300%22%20height%3D%22150%22%20viewBox%3D%220%200%20300%20150%22%3E%3Crect%20fill%3D%22%23ddd%22%20width%3D%22300%22%20height%3D%22150%22%3E%3C%2Frect%3E%3Ctext%20fill%3D%22rgba%280%2C0%2C0%2C0.5%29%22%20font-family%3D%22sans-serif%22%20font-size%3D%2230%22%20font-weight%3D%22bold%22%20dy%3D%2210.5%22%20x%3D%2250%25%22%20y%3D%2250%25%22%20text-anchor%3D%22middle%22%3E300x150%3C%2Ftext%3E%3C%2Fsvg%3E', $this->extension->getSvgPlaceholderUri());
    }
}


src/SvgPlaceholderExtension.php
<?php

declare(strict_types=1);

namespace Kerox\TwigImagePlaceholder;

use Twig\Extension\AbstractExtension;
use Twig\TwigFunction;

final class SvgPlaceholderExtension extends AbstractExtension
{
    private const DEFAULT_OPTIONS = [
        'fontFamily' => 'sans-serif',
        'fontWeight' => 'bold',
        'bgColor' => '#ddd',
        'textColor' => 'rgba(0,0,0,0.5)',
    ];

    public function getFunctions(): array
    {
        return [
            new TwigFunction('svg_placeholder', [$this, 'getSvgPlaceholder'], ['is_safe' => ['html']]),
            new TwigFunction('svg_placeholder_uri', [$this, 'getSvgPlaceholderUri']),
        ];
    }

    public function getSvgPlaceholder(int $width = 300, int $height = 150, array $options = []): string
    {
        $options = array_replace(self::DEFAULT_OPTIONS, $options);
        $options += [
            'fontSize' => $fontSize = floor(min($width, $height) * 0.2),
            'dy' => $fontSize * 0.35,
            'text' => sprintf('%dx%d', $width, $height),
        ];

        $doc = new \DOMDocument('1.0');

        // svg element
        $svgElement = $doc->createElement('svg');
        $svgElement->setAttribute('xmlns', 'http://www.w3.org/2000/svg');
        $svgElement->setAttribute('width', (string) $width);
        $svgElement->setAttribute('height', (string) $height);
        $svgElement->setAttribute('viewBox', "0 0 $width $height");
        if ($class = $options['class'] ?? null) {
            $svgElement->setAttribute('class', $class);
        }

        // rect element
        $rectElement = $doc->createElement('rect');
        $rectElement->setAttribute('fill', $options['bgColor']);
        $rectElement->setAttribute('width', (string) $width);
        $rectElement->setAttribute('height', (string) $height);

        $svgElement->appendChild($rectElement);

        // text element
        if ($text = $options['text'] ?? null) {
            $textElement = $doc->createElement('text');
            $textElement->setAttribute('fill', $options['textColor']);
            $textElement->setAttribute('font-family', $options['fontFamily']);
            $textElement->setAttribute('font-size', (string) $options['fontSize']);
            $textElement->setAttribute('font-weight', (string) $options['fontWeight']);
            $textElement->setAttribute('dy', (string) $options['dy']);
            $textElement->setAttribute('x', '50%');
            $textElement->setAttribute('y', '50%');
            $textElement->setAttribute('text-anchor', 'middle');
            $textElement->appendChild($doc->createTextNode($text));

            $svgElement->appendChild($textElement);
        }

        $doc->appendChild($svgElement);

        return trim($doc->saveHTML());
    }

    public function getSvgPlaceholderUri(int $width = 300, int $height = 150, array $options = []): string
    {
        return sprintf('data:image/svg+xml;charset=UTF-8,%s', rawurlencode($this->getSvgPlaceholder($width, $height, $options)));
    }
}
